python/ray/experimental/sgd_api.py

```diff
@@ -0,0 +1,315 @@
+        Returns:
+            Deserialized return values from the target function.
+        """
+        return ray.get([worker.execute.remote(fn) for worker in self.workers])
+
+    def run(self,
+            fn: Callable[[Any], Any],
+            *args: Optional[List] = None,
+            **kwargs: Optional[Dict] = None) -> List[Any]:
+        """Executes the provided function on all workers.
+        Args:
+            fn: Target function that can be executed with arbitrary
+                args and keyword arguments.
+            args: List of arguments to be passed into the target function.
+            kwargs: Dictionary of keyword arguments to be
+                passed into the target function.
+        Returns:
+            Deserialized return values from the target function.
+        """
+        return ray.get(self.run_remote(fn, args, kwargs))
+
+    def run_remote(self,
+                   fn: Callable[[Any], Any],
+                   *args: Optional[List] = None,
+                   **kwargs: Optional[Dict] = None) -> List[Any]:
+        """Executes the provided function on all workers.
+        Args:
+            fn: Target function that can be executed with arbitrary
+                args and keyword arguments.
+            args: List of arguments to be passed into the target function.
+            kwargs: Dictionary of keyword arguments to be
+                passed into the target function.
+        Returns:
+            list: List of ObjectRefs that you can run `ray.get` on to
+                retrieve values.
+        """
+        return [
+            worker.execute.remote(lambda w: fn(args, kwargs))
+            for worker in self.workers
+        ]
+
+    def execute_single(self,
+                       fn: Callable[["executable_cls"], Any]) -> List[Any]:
+        """Executes the provided function on the rank 0 worker (chief).
+        Args:
+            fn: Target function to be invoked on the chief object.
+        Returns:
+            Deserialized return values from the target function.
+        """
+        return ray.get(self.workers[0].execute.remote(fn))
+
+    def set_env_vars(env_vars: Dict):
+        ray.get(w.set_env_vars.remote(env_vars) for w in self.workers]
+
+
+    def shutdown(self):
+        """Destroys the provided workers."""
+        for worker in self.workers:
+            del worker
+
+        remove_placement_group(self.pg)
+
+
+# usage
+class Foo:
+    def __init__(self, counter):
+        self.counter = counter
+
+    def increment(self):
+        self.counter += 1
+
+actor_group = ActorGroup(num_workers=2)
+actor_group.start(Foo, counter=0)
+for _ in range(5):
+    actor_group.run(lambda w: w.increment())
+
+assert actor_group.run(lambda w: w.counter) == [5, 5]
+
+
+# SGD Backends.
+
+# Framework specific actor groups. Allows you to do framework specific
+# implementations
+
+# Torch: Setup process groups
+# Horovod: Create Coordinator, NIC detection, ...
+
+class TorchActorGroup(ActorGroup):
```

Comment:
nit, not sure what the benefit of subclassing here is? it seems like you just have some functionality to run after `start`. You can achieve the same with:

```
x = actor_group()
x.run(lambda _: setup_torch_process())

Reply:
Yep agree here, subclassing is not necessary